sdk/dotnet/Pulumi/Deployment/Deployment.Runner.cs

```diff
@@ -80,21 +84,31 @@ Task<int> IRunner.RunAsync(Func<Task<IDictionary<string, object?>>> func, StackO
             public void RegisterTask(string description, Task task)
             {
                 _deploymentLogger.LogDebug($"Registering task: {description}");
+                _inFlightTasks.AddTask(task);
 
-                lock (_inFlightTasks)
+                // Ensure completion message is logged at most once when the task finishes.
+                if (_deploymentLogger.IsEnabled(LogLevel.Debug))
                 {
                     // We may get several of the same tasks with different descriptions.  That can
                     // happen when the runtime reuses cached tasks that it knows are value-identical
                     // (for example Task.CompletedTask).  In that case, we just store all the
                     // descriptions. We'll print them all out as done once this task actually
                     // finishes.
-                    if (!_inFlightTasks.TryGetValue(task, out var descriptions))
+
+                    var key = new Tuple<int,string>(task.Id, description);
+                    int timesSeen = _descriptions.AddOrUpdate(
+                        key,
+                        (Tuple<int,string> key) => 1,
+                        (Tuple<int,string> key, int value) => value + 1
+                    );
+                    if (timesSeen == 1)
                     {
-                        descriptions = new List<string>();
-                        _inFlightTasks.Add(task, descriptions);
+                        task.ContinueWith(task => {
+                            _deploymentLogger.LogDebug($"Completed task: {description}");
+                            int old;
+                            _descriptions.TryRemove(key, out old);
```

Comment:
You can declare the out variable inline to save yourself having to declare it early:
```csharp
_descriptions.TryRemove(key, out var old);
```

But since here we don't need the out variable you can omit the declaration altogether and use the discard:
```suggestion
                            _descriptions.TryRemove(key, out _);
```